packages/http-client/src/index.ts

```diff
@@ -0,0 +1,202 @@
+import { effect as T } from "@matechs/effect";
+import { ParsedUrlQueryInput } from "querystring";
+
+/* tested in the implementation packages */
+/* istanbul ignore file */
+
+export enum Method {
+  GET,
+  POST,
+  PUT,
+  DELETE,
+  PATCH
+}
+
+export enum RequestType {
+  JSON,
+  DATA
+}
+
+export type Headers = Record<string, string>;
+
+export interface Response<Body> {
+  body?: Body;
+  headers: Headers;
+  status: number;
+}
+
+export interface HttpResponseError<ErrorBody> {
+  _tag: HttpErrorReason.Response;
+  response: Response<ErrorBody>;
+}
+
+export interface HttpRequestError {
+  _tag: HttpErrorReason.Request;
+  error: Error;
+}
+
+export interface HttpDeserializer {
+  httpDeserializer: {
+    response: <A>(a: string) => A | undefined;
+    errorResponse: <E>(error: string) => E | undefined;
+  };
+}
+
+export enum HttpErrorReason {
+  Request,
+  Response
+}
+
+export type HttpError<ErrorBody> =
+  | HttpRequestError
+  | HttpResponseError<ErrorBody>;
+
+export function foldHttpError<A, B, ErrorBody>(
+  onError: (e: Error) => A,
+  onResponseError: (e: Response<ErrorBody>) => B
+): (err: HttpError<ErrorBody>) => A | B {
+  return err => {
+    switch (err._tag) {
+      case HttpErrorReason.Request:
+        return onError(err.error);
+      case HttpErrorReason.Response:
+        return onResponseError(err.response);
+    }
+  };
+}
+
+export interface HttpHeaders {
+  headers: Record<string, string>;
+}
+
+export interface Http {
+  http: {
+    request: <I, E, O>(
+      method: Method,
+      url: string,
+      headers: Record<string, string>,
+      body?: I,
+      requestType?: RequestType
+    ) => T.Effect<HttpDeserializer, HttpError<E>, Response<O>>;
+  };
+}
+
+function hasHeaders(r: any): r is HttpHeaders {
```

Comment:
This doesn't look type safe, if this function could actually take a parameter of any type, consider this type safe implementation:
```typescript
function hasHeaders(r: unknown): r is HttpHeaders {
  return typeof r === 'object' && !!r && "headers" in r;
}
```

Reply:
this takes an environment that is always defined, worst case is {}

Reply:
still I prefer your signature.